### PR TITLE
feat: Geospatial Data Type and GIS Function Support for milvus server

### PR DESCRIPTION
**Pull request by tasty-gumi**
issue:https://github.com/milvus-io/milvus/issues/27576

# Main Goals
1. Create and describe collections with geospatial fields, enabling both client and server to recognize and process geo fields.
2. Insert geospatial data as payload values in the insert binlog, and print the values for verification.
3. Load segments containing geospatial data into memory.
4. Ensure query outputs can display geospatial data.
5. Support filtering on GIS functions for geospatial columns.

# Solution
1. **Add Type**: Modify the Milvus core by adding a Geospatial type in both the C++ and Go code layers, defining the Geospatial data structure and the corresponding interfaces.
2. **Dependency Libraries**: Introduce necessary geospatial data processing libraries. In the C++ source code, use Conan package management to include the GDAL library. In the Go source code, add the go-geom library to the go.mod file.
3. **Protocol Interface**: Revise the Milvus protocol to provide mechanisms for Geospatial message serialization and deserialization.
4. **Data Pipeline**: Facilitate interaction between the client and proxy using the WKT format for geospatial data. The proxy will convert all data into WKB format for downstream processing, providing column data interfaces, segment encapsulation, segment loading, payload writing, and cache block management.
5. **Query Operators**: Implement simple display and support for filter queries. Initially, focus on filtering based on spatial relationships for a single column of geospatial literal values, providing parsing and execution for query expressions.
6. **Index Construction**: Consider building an H3 index, utilizing the C interface provided by the H3 system.
7. **Client Modification**: Enable the client to handle user input for geospatial data and facilitate end-to-end testing.Check the modification in pymilvus.

 delete incomplete H3 Index development and useless generated files.

### CODE REVIEW

**Comment by czs007** [comment]
issue:https://github.com/milvus-io/milvus/issues/27576

File: internal/core/src/common/GeoSpatial.h [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            wkb_data_ = new unsigned char[size_];
+            // little-endian order to save wkb
+            geometry_->exportToWkb(wkbNDR, wkb_data_);
+        }
+    }
+
+    //read write ptr, use to save a OGRGeometry object when need to create new storage file
+    unsigned char* wkb_data_{nullptr};
+    size_t size_{0};
+    OGRGeometry* geometry_{nullptr};
```
Comment (alexanderguzhva) [inline internal/core/src/common/GeoSpatial.h:153]: can `std::unique_ptr<OGRGeometry>` be used here instead of a raw pointer?

File: internal/core/src/common/GeoSpatial.h [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        if (geometry_ && size_ == 0) {
+            size_ = geometry_->WkbSize();
+            wkb_data_ = new unsigned char[size_];
+            // little-endian order to save wkb
+            geometry_->exportToWkb(wkbNDR, wkb_data_);
+        }
+    }
+
+    //read write ptr, use to save a OGRGeometry object when need to create new storage file
+    unsigned char* wkb_data_{nullptr};
```
Comment (alexanderguzhva) [inline internal/core/src/common/GeoSpatial.h:151]: can `std::unique_ptr<unsigned char[]>` be used here instead of a raw pointer?

Comment (tasty-gumi) [reply internal/core/src/common/GeoSpatial.h:151]: Thanks for your suggestions.

File: configs/milvus.yaml [hunk 3]
```diff
@@ -85,10 +85,7 @@ tikv:
     tlsCACert:  # path to your CACert file
 
 localStorage:
-  # Local path to where vector data are stored during a search or a query to avoid repetitve access to MinIO or S3 service.
-  # Caution: Changing this parameter after using Milvus for a period of time will affect your access to old data.
-  # It is recommended to change this parameter before starting Milvus for the first time.
-  path: /var/lib/milvus/data/
+  path: /tmp/milvus/data/ # please adjust in embedded Milvus: /tmp/milvus/data/
```
Comment (czs007) [inline configs/milvus.yaml:88]: revert this

Comment (czs007) [reply configs/milvus.yaml:88]: revert all changes in this file.

File: deployments/docker/dev/docker-compose.yml [hunk 4]
```diff
@@ -49,13 +49,7 @@ services:
       - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/minio:/minio_data
     command: minio server /minio_data --console-address ":9001"
     healthcheck:
```
Comment (czs007) [inline deployments/docker/dev/docker-compose.yml:51]: revert all changes in this file

File: .gitignore [hunk 5]
```diff
@@ -105,3 +105,7 @@ internal/core/src/pb/*.pb.h
 internal/core/src/pb/*.pb.cc
 *.pb.go
 **/legacypb/*.pb.go
+
```
Comment (czs007) [inline .gitignore:108]: revert all changes in this file.

File: go.mod [hunk 6]
```diff
@@ -274,3 +275,5 @@ replace (
 )
 
 exclude github.com/apache/pulsar-client-go/oauth2 v0.0.0-20211108044248-fe3b7c4e445b
+
+replace github.com/milvus-io/milvus-proto/go-api/v2 => /home/adam/milvus-proto/go-api
```
Comment (czs007) [inline go.mod:279]: revert this change.

File: internal/core/src/common/FieldData.cpp [hunk 7]
```diff
@@ -206,9 +206,16 @@ FieldDataImpl<Type, is_type_entire_row>::FillFieldData(
                 values[index] =
                     Json(simdjson::padded_string(json_array->GetString(index)));
             }
-            if (nullable_) {
-                return FillFieldData(
-                    values.data(), array->null_bitmap_data(), element_count);
```
Comment (czs007) [inline internal/core/src/common/FieldData.cpp:211]: revert this change

File: internal/core/src/storage/Event.cpp [hunk 8]
```diff
@@ -282,6 +283,17 @@ BaseEventData::Serialize() {
             }
             break;
         }
+        case DataType::GEOSPATIAL: {
+            for (size_t offset = 0; offset < field_data->get_num_rows();
+                 ++offset) {
+                auto geo_ptr = static_cast<const std::string*>(
+                    field_data->RawValue(offset));
+                payload_writer->add_one_binary_payload(
```
Comment (smellthemoon) [inline internal/core/src/storage/Event.cpp:291]: change it to the size to" auto size =
                    field_data->is_valid(offset) ? string_view.size() : -1;", which can help to add null in payload.

**Comment by smellthemoon** [comment]
please add some test in payload_test.go to ensure accurate reading and writing on the go side.

File: internal/core/src/expr/ITypeExpr.h [hunk 9]
```diff
@@ -705,6 +706,34 @@ class CompareExpr : public ITypeFilterExpr {
...
+        : column_(cloumn), op_(op), wkb_(wkb){};
+    std::string
+    ToString() const override {
+        OGRGeometry* geo = nullptr;
+        OGRGeometryFactory::createFromWkb(
+            wkb_.data(), nullptr, &geo, wkb_.size());
+        if (!geo) {
+            return "";
+        }
+        return fmt::format(
```
Comment (czs007) [inline internal/core/src/expr/ITypeExpr.h:723]: geo should be freed with [OGRGeometryFactory::destroyGeometry()](https://gdal.org/en/latest/doxygen/classOGRGeometryFactory.html#aae001086e26985d95c36ccd255a8c6d5) after use.
[OGRGeometryFactory::destroyGeometry()](https://gdal.org/en/latest/doxygen/classOGRGeometryFactory.html#aae001086e26985d95c36ccd255a8c6d5)

File: internal/core/src/mmap/Utils.h [hunk 10]
```diff
@@ -136,6 +137,27 @@ WriteFieldData(File& file,
                 }
                 break;
             }
+            case DataType::GEOSPATIAL: {
+                // write as: |size|data|size|data......
+                for (ssize_t i = 0; i < data->get_num_rows(); ++i) {
+                    indices.push_back(total_written);
+                    auto Geo_ptr =
```
Comment (czs007) [inline internal/core/src/mmap/Utils.h:144]: The variable name "Geo_ptr" is not appropriately standardized.

File: internal/core/src/query/PlanProto.cpp [hunk 11]
```diff
@@ -384,6 +386,25 @@ ProtoParser::ParseValueExprs(const proto::plan::ValueExpr& expr_pb) {
...
 
+expr::TypedExprPtr
+ProtoParser::ParseGISFunctionFilterExprs(
+    const proto::plan::GISFunctionFilterExpr& expr_pb) {
+    auto& columnInfo = expr_pb.column_info();
+    auto field_id = FieldId(columnInfo.field_id());
+    auto data_type = schema[field_id].get_data_type();
+    Assert(data_type == (DataType)columnInfo.data_type());
+    const std::string& str = expr_pb.wkt_string();
+    OGRGeometry* geometry = nullptr;
```
Comment (czs007) [inline internal/core/src/query/PlanProto.cpp:397]: remember to destroy geometry

File: internal/core/src/query/PlanProto.cpp [hunk 12]
```diff
@@ -384,6 +386,25 @@ ProtoParser::ParseValueExprs(const proto::plan::ValueExpr& expr_pb) {
...
+    OGRGeometryFactory::createFromWkt(str.data(), nullptr, &geometry);
+    Assert(geometry != nullptr);
+    unsigned char* wkb_byte = new unsigned char[geometry->WkbSize()];
```
Comment (czs007) [inline internal/core/src/query/PlanProto.cpp:400]: wkb_type is leaked

File: internal/parser/planparserv2/Plan.g4 [hunk 13]
```diff
@@ -87,6 +94,14 @@ ArrayContainsAll: 'array_contains_all' | 'ARRAY_CONTAINS_ALL';
 ArrayContainsAny: 'array_contains_any' | 'ARRAY_CONTAINS_ANY';
 ArrayLength: 'array_length' | 'ARRAY_LENGTH';
 
+GeospatialEuqals:'geospatial_equals' | 'GEOSPATIAL_EQUALS';
```
Comment (czs007) [inline internal/parser/planparserv2/Plan.g4:97]: rename to st_equals and ST_EQUALS

File: internal/proto/plan.proto [hunk 14]
```diff
@@ -139,6 +139,22 @@ message JSONContainsExpr {
   bool elements_same_type = 4;
 }
 
+message GISFunctionFilterExpr{
```
Comment (czs007) [inline internal/proto/plan.proto:142]: TODO: use custom filtering function in next pr.

File: tests/integration/import/import_test.go [hunk 15]
```diff
@@ -246,10 +266,10 @@ func (s *BulkInsertSuite) TestMultiFileTypes() {
 
 		// TODO: not support numpy for SparseFloatVector by now
 		if fileType != importutilv2.Numpy {
-			s.vecType = schemapb.DataType_SparseFloatVector
-			s.indexType = indexparamcheck.IndexSparseWand
-			s.metricType = metric.IP
-			s.run()
+			// s.vecType = schemapb.DataType_SparseFloatVector
```
Comment (czs007) [inline tests/integration/import/import_test.go:269]: why comment this?

File: scripts/generate_proto.sh [hunk 16]
```diff
@@ -25,7 +25,8 @@ done
 ROOT_DIR="$( cd -P "$( dirname "$SOURCE" )/.." && pwd )"
 
 PROTO_DIR=$ROOT_DIR/internal/proto/
-API_PROTO_DIR=$ROOT_DIR/cmake_build/thirdparty/milvus-proto/proto
+# API_PROTO_DIR=$ROOT_DIR/cmake_build/thirdparty/milvus-proto/proto
+API_PROTO_DIR=/home/adam/milvus-proto/proto
```
Comment (czs007) [inline scripts/generate_proto.sh:29]: revert this after https://github.com/milvus-io/milvus-proto/pull/320 is merged.

File: pkg/go.mod [hunk 17]
```diff
@@ -197,3 +198,5 @@ replace (
 )
 
 exclude github.com/apache/pulsar-client-go/oauth2 v0.0.0-20211108044248-fe3b7c4e445b
+
+replace github.com/milvus-io/milvus-proto/go-api/v2 => /home/adam/milvus-proto/go-api
```
Comment (czs007) [inline pkg/go.mod:202]: revert this after milvus-proto pr merged.

File: internal/parser/planparserv2/parser_visitor.go [hunk 18]
```diff
@@ -1370,3 +1370,213 @@ func (v *ParserVisitor) VisitArrayLength(ctx *parser.ArrayLengthContext) interfa
...
+	columnInfo, err := v.getChildColumnInfo(ctx.Identifier(), ctx.StringLiteral())
+	if err != nil {
+		return err
+	}
+	if columnInfo == nil ||
+		(!typeutil.IsGeospatialType(columnInfo.GetDataType())) {
+		return fmt.Errorf(
+			"Euqals operation are only supported on geospatial fields now, got: %s", ctx.GetText())
+	}
+	element := ctx.GetChild(4).(antlr.ParseTree).GetText() // the wkt input
```
Comment (xiaocai2333) [inline internal/parser/planparserv2/parser_visitor.go:1384]: elements, ok := ctx.GetChild(4).(antlr.ParseTree).GetText()
add `ok` to ensure it won't panic

File: internal/parser/planparserv2/parser_visitor.go [hunk 19]
```diff
@@ -1370,3 +1370,213 @@ func (v *ParserVisitor) VisitArrayLength(ctx *parser.ArrayLengthContext) interfa
...
```
Comment (xiaocai2333) [inline internal/parser/planparserv2/parser_visitor.go:1414]: ditto

File: internal/util/hookutil/mock_hook.go [hunk 20]
```diff
@@ -1,6 +1,3 @@
-//go:build test
-// +build test
-
 /*
  * Licensed to the LF AI & Data foundation under one
  * or more contributor license agreements. See the NOTICE file
```
Comment (czs007) [inline internal/util/hookutil/mock_hook.go:3]: revert this change.

File: internal/parser/planparserv2/parser_visitor.go (hunk 18 above)
Comment (xiaocai2333) [reply internal/parser/planparserv2/parser_visitor.go:1384]: and why get `4` for ctx child? what's the meaning?

File: internal/parser/planparserv2/parser_visitor.go [hunk 21]
```diff
@@ -1370,3 +1370,213 @@ func (v *ParserVisitor) VisitArrayLength(ctx *parser.ArrayLengthContext) interfa
 		nodeDependent: true,
 	}
 }
+
+func (v *ParserVisitor) VisitGeospatialEuqals(ctx *parser.GeospatialEuqalsContext) interface{} {
+	columnInfo, err := v.getChildColumnInfo(ctx.Identifier(), ctx.StringLiteral())
```
Comment (czs007) [inline internal/parser/planparserv2/parser_visitor.go:1375]: no need to pass parameter `ctx.StringLiteral`

Comment (czs007) [reply internal/parser/planparserv2/parser_visitor.go:1375]: refer 

childExpr, err := v.translateIdentifier(identifier.GetText())
		if err != nil {
			return nil, err
		}
		return toColumnInfo(childExpr), nil

File: internal/core/src/common/Geometry.h [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+            size_ = geometry_->WkbSize();
+            wkb_data_ = std::make_unique<unsigned char[]>(size_);
+            // little-endian order to save wkb
+            geometry_->exportToWkb(wkbNDR, wkb_data_.get());
+        }
+    }
+
+    std::unique_ptr<unsigned char[]> wkb_data_;
+    size_t size_{0};
+    std::unique_ptr<OGRGeometry> geometry_;
```
Comment (czs007) [inline internal/core/src/common/Geometry.h:169]: how to free geometry_ ?

File: internal/core/src/common/Geometry.h [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+
+    std::string
+    to_wkb_string() const {
+        return std::string(reinterpret_cast<char*>(wkb_data_.get()), size_);
+    }
+
+ private:
+    inline void
+    to_wkb_internal() {
+        if (geometry_ && size_ == 0) {
```
Comment (czs007) [inline internal/core/src/common/Geometry.h:159]: why check size_ == 0 ?

File: internal/core/src/query/PlanProto.cpp [hunk 24]
```diff
@@ -221,6 +226,7 @@ ProtoParser::CreateRetrievePlan(const proto::plan::PlanNode& plan_node_proto) {
 
     auto plan_node = RetrievePlanNodeFromProto(plan_node_proto);
```
Comment (jiangyinzuo) [inline internal/core/src/query/PlanProto.cpp:228]: forget to remove `LOG_WARN(...)` 🤔

File: internal/core/src/query/PlanProto.cpp [hunk 25]
```diff
@@ -221,6 +226,7 @@ ProtoParser::CreateRetrievePlan(const proto::plan::PlanNode& plan_node_proto) {
...
 
+    LOG_WARN("CREATED PALNNODE!!");
```
Comment (jiangyinzuo) [inline internal/core/src/query/PlanProto.cpp:229]: do not forget to remove `LOG_WARN(...)`

**Comment by tasty-gumi** [comment]
[make-milvus : make] gdal/3.5.3: Retrieving from server 'default-conan-local' 

[make-milvus : make] gdal/3.5.3: Trying with 'default-conan-local'...

[make-milvus : make] Downloading conanmanifest.txt

[make-milvus : make] Downloading conanfile.py

[make-milvus : make] gdal/3.5.3: Downloaded recipe revision 0d131926ca2d68ab6c132e045a0e097d

[make-milvus : make] ERROR: gdal/3.5.3: option 'shared' doesn't exist

[make-milvus : make] Possible options are []

[make-milvus : make] conan install failed

[make-milvus : make] make: *** [Makefile:251: build-3rdparty] Error 1

[make-milvus : make] ldd: ./bin/milvus: No such file or directory

[make-milvus : make] /milvus/lib/

**Comment by tasty-gumi** [comment]
when I run conan get gdal/3.5.3@ -r default-conan-local
the output has this option:
 import os

from conan import ConanFile
from conan.errors import ConanInvalidConfiguration
from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
from conan.tools.files import apply_conandata_patches, copy, export_conandata_patches, get, rmdir, replace_in_file
from conan.tools.microsoft import is_msvc
from conan.tools.scm import Version

required_conan_version = ">=1.52.0"


class GdalConan(ConanFile):
    name = "gdal"
    description = "GDAL is an open source X/MIT licensed translator library " \
                  "for raster and vector geospatial data formats."
    license = "MIT"
    url = "https://github.com/conan-io/conan-center-index"
    homepage = "https://github.com/OSGeo/gdal"
    topics = ("osgeo", "geospatial", "raster", "vector")

    package_type = "library"
    settings = "os", "arch", "compiler", "build_type"
    options = {
        "shared": [True, False],
        "fPIC": [True, False],
        "tools": [True, False],
        
        ....
